tests/checker/test_https.py

```diff
@@ -15,34 +15,50 @@
 # with this program; if not, write to the Free Software Foundation, Inc.,
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 """
-Test news checking.
+Test https.
 """
-import pytest
+import subprocess
 
 from tests import need_network
-from . import LinkCheckTest
+from .httpserver import HttpsServerTest, CookieRedirectHttpRequestHandler
+from .. import get_file
 
+from linkcheck import httputil
 
-class TestHttps (LinkCheckTest):
+from OpenSSL import crypto
+
+class TestHttps(HttpsServerTest):
     """
     Test https: link checking.
     """
 
-    @pytest.mark.xfail
+    def __init__(self, methodName='runTest'):
+        super(TestHttps, self).__init__(methodName=methodName)
+        self.handler = CookieRedirectHttpRequestHandler
+        temp_key_file = get_file("key.pem")
+        subprocess.check_output(["openssl", "req", "-x509", "-newkey",
+            "rsa:2048", "-passout", "pass:linkchecker", "-subj", "/DC=org",
+            "-keyout", temp_key_file, "-out", get_file("https_cert.pem")])
+        # Strip pass phrase from the key, else it is requested at server start
+        subprocess.check_output(["openssl", "pkey",
+            "-passin", "pass:linkchecker",
+            "-in", temp_key_file, "-out", get_file("https_key.pem")])
```

Comment:
1. I think this belongs in `setUp`, not `__init__`
2. Surely there's a way of generating a self-signed cert without a password
3. Generating a new SSL key/cert for every test seems expensive, can we not do that when the key already exists?  (And maybe set expiration date to something long -- a decade, a thousand years -- to prevent problems from cert reuse.)
4. I wonder if pyopenssl (or cryptogtaphy) have APIs for generating SSL keys/self-signed certs, ti avoid shelling out.

Reply:
2. I believe the `-nodes` argument to `openssl req` suppresses the password prompt (it means "don't encrypt the key" because openssl's command-line UX is terrible).

Reply:
Re 4: https://skippylovesmalorie.wordpress.com/2010/02/12/how-to-generate-a-self-signed-certificate-using-pyopenssl/

Reply:
> 3. Generating a new SSL key/cert for every test seems expensive, can we not do that when the key already exists? (And maybe set expiration date to something long -- a decade, a thousand years -- to prevent problems from cert reuse.)

There should be a way to generate SSL keys with garbage entropy or `/dev/urandom` so that it's fast. There's such a flag in GnuPG for sure, so maybe there's something similar in openssl?

Seems like a reasonable request anyways.

Reply:
>     1. I think this belongs in `setUp`, not `__init__`

Is `setUp` called for each method, the docs aren't really clear? I did think about `setUpClass`, saw New in 3.2 and didn't use it - however it is in 2.7 and seems to work.

Reply:
> Re 4: https://skippylovesmalorie.wordpress.com/2010/02/12/how-to-generate-a-self-signed-certificate-using-pyopenssl/

I will have a go. That link suggests you can set the notAfter time which could make for a better test than just True.

Reply:
Yes, setUp is called before each test method.

Reply:
Note that `__init__` is called once for every test method too, while constructing the test suite!

The unittest API requires that each instance of a test case has a runTest() method that runs the test.  When you define a test class, the default test loader invokes unittest.MakeSuite(YourTestClass) which creates an empty TestSuite(), enumerates all the test methods of the class, and then instantiates the class by passing the test method name to `__init__` to get a test case instance that will run that particular method from its runTest().

So if you wanted to avoid creating new SSL keys for every test method, `__init__ `is the wrong way to go about it.

(Caveat: I don't know for sure that pytest's unittest compatibility mode also works the same way, but I think it's very likely.)